mlflow/sklearn/__init__.py

```diff
@@ -894,12 +1263,104 @@ def patched_fit(original, self, *args, **kwargs):
                           `sklearn.linear_model.LogisticRegression.fit()` is being patched)
         """
+        status = _get_autologging_metrics_manager()
         with _SklearnTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                # In `fit_mlflow` call, it will also call metric API for computing training metrics
+                # so we need temporarily disable the post_training_metrics patching.
+                with status.disable_log_post_training_metrics():
+                    result = fit_mlflow(original, self, *args, **kwargs)
+                status.register_model(self, mlflow.active_run().info.run_id)
+                return result
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_predict(original, self, *args, **kwargs):
+        status = _get_autologging_metrics_manager()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            predict_result = original(self, *args, **kwargs)
+            eval_dataset = get_instance_method_first_arg_value(original, args, kwargs)
+            eval_dataset_name = status.register_prediction_input_dataset(self, eval_dataset)
+
+            """
+            In `patched_predict`, register the prediction result instance with the run id and
+             eval dataset name. e.g.
+    
+            ```
+            prediction_result = model_1.predict(eval_X)
+            ```
+            then we need register the following relatinoship into the
+            `_autologging_metrics_manager`:
+            id(prediction_result) --> (eval_dataset_name, run_id)
+            
+            Note: we cannot set additional attributes "eval_dataset_name" and "run_id" into
+            the prediction_result object, because certain dataset type like numpy does not support
+            additional attribute assignment.
+            """
+            status.register_prediction_result(
+                status.get_run_id_for_model(self),
+                eval_dataset_name,
+                predict_result
+            )
+            return predict_result
+        else:
+            return original(self, *args, **kwargs)
+
+    def patched_metric_api(original, *args, **kwargs):
+        status = _get_autologging_metrics_manager()
+        if status.should_log_post_training_metrics():
+            # one metric api may call another metric api,
+            # to avoid this, call disable_log_post_training_metrics to avoid nested patch
+            with status.disable_log_post_training_metrics():
+                metric = original(*args, **kwargs)
+
+            if status.is_metrics_value_loggable(metric):
+                metric_name = original.__name__
+                call_command = status.gen_metric_call_command(None, original, *args, **kwargs)
```

Comment:
Nit:

Can we just define `gen_metric_call_command` as follows:

```python
def gen_metric_call_command(..., args, kwargs):
   ...
```

and call it as follows:

```python
gen_metric_call_command(..., args, kwargs)
```

Reply:
No. The `*args, **kwargs` is intentional. Because in `gen_metric_call_command`, it will call `inspect_variable_name` method, the inspecting variable name will tracing variable name to the outmost call frame **via each frame argument**. If defined arguments like `args, kwargs`, then the true argument will be hidden inside the `args` list and `kwargs` dict, which block the `inspect_variable_name` to trace the original variable in outmost call frame.

Reply:
Got it, thanks for the explanation!